setup.sh

```diff
@@ -104,3 +104,10 @@ echo "${RUNNER_TOOL_CACHE}/dart-sdk/bin" >> $GITHUB_PATH
 # Report success, and print version.
 echo -e "Successfully installed Dart SDK:"
 ${RUNNER_TOOL_CACHE}/dart-sdk/bin/dart --version
+
+# When enabled through env variables, create OIDC token for publishing on pub.dev.
+if [[ "${ACTIONS_ID_TOKEN_REQUEST_URL}" != "" && "${ACTIONS_ID_TOKEN_REQUEST_TOKEN}" != "" ]]
+then
+  PUB_TOKEN=$(curl -sLS "${ACTIONS_ID_TOKEN_REQUEST_URL}&audience=https://pub.dev" -H "User-Agent: actions/oidc-client" -H "Authorization: Bearer $ACTIONS_ID_TOKEN_REQUEST_TOKEN" | jq -r '.value')
```

Comment:
Maybe, do `--retry 5 --retry-connrefused` for good measure.

Reply:
Also won't we need to do `export PUB_TOKEN`.

Also will this actually work... the command `dart pub token add https://pub.dev --env-var PUB_TOKEN` will read from environment variable `PUB_TOKEN` when making the requests (not when setting the value).

Reply:
We could rename it to `PUB_DEV_PUBLISHING_TOKEN` to reduce name collisions or unintended overrides, but I don't think we need to care much about intentional overrides or misuses.

Reply:
I'm still not sure this will work. The command `dart pub token add https://pub.dev --env-var PUB_TOKEN` will **not** read the `PUB_TOKEN` env var.

This will only be done by the `dart pub publish` command, which is called in user-code. Which probably isn't sourcing in the `setup.sh` script.

So maybe we should:
 * Test if this works?
 * Try: `curl -sLS "${ACTIONS_ID_TOKEN_REQUEST_URL}&audience=https://pub.dev" -H "User-Agent: actions/oidc-client" -H "Authorization: Bearer $ACTIONS_ID_TOKEN_REQUEST_TOKEN" | jq -r '.value' | dart pub add https://pub.dev`
 * Consider if we need a cleanup step that removes the config file after running the script.

Reply:
FYI: Environment variables can be made available to subsequent steps using the file pointed to by `$GITHUB_ENV`: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable

```sh
echo "PUB_TOKEN=$(curl --retry 5 --retry-connrefused -sLS "${ACTIONS_ID_TOKEN_REQUEST_URL}&audience=https://pub.dev" -H "User-Agent: actions/oidc-client" -H "Authorization: Bearer $ACTIONS_ID_TOKEN_REQUEST_TOKEN" | jq -r '.value')" >> $GITHUB_ENV
```

Regarding writing the private token to a file: [google-github-actions/auth](https://github.com/google-github-actions/auth) does that as well for GCP Service Account credentials:

> `create_credentials_file`: (Optional) If true, the action will securely
> generate a credentials file which can be used for authentication via gcloud
> and Google Cloud SDKs in other steps in the workflow. The default is true.
>
> The credentials file is exported into `$GITHUB_WORKSPACE`, which makes it
> available to all future steps and filesystems (including Docker-based
> GitHub Actions). The file is automatically removed at the end of the job
> via a post action. In order to use exported credentials, you **must** add
> the `actions/checkout` step before calling `auth`. This is due to how
> GitHub Actions creates `$GITHUB_WORKSPACE`:
>
> ```yaml
> jobs:
> job_id:
>   steps:
>   - uses: 'actions/checkout@v3' # Must come first!
>   - uses: 'google-github-actions/auth@v0'
> ```

After that, an environment variable `GOOGLE_APPLICATION_CREDENTIALS` pointing to this newly created file is created for subsequent steps.

The post action implementation removing this credentials file can be found here: https://github.com/google-github-actions/auth/blob/main/src/post.ts